### ISSUE TITLE
Changes to unbind causes null entries (and that leads to bad bindings)

### ISSUE THREAD

**Issue by Vaccano** [aurelia/templating#310]
This pull request created by @jdanyow  https://github.com/aurelia/templating/pull/306 fixed a bug that was blocking me, so I added it to my application like this:

```
import {View} from 'aurelia-templating';

View.prototype.unbind = function unbind() {
    var controllers = void 0;
    var bindings = void 0;
    var children = void 0;
    var i = void 0;
    var ii = void 0;

    if (this.isBound) {
        this.isBound = false;
        this.resources._invokeHook('beforeUnbind', this);
        
        if (this.controller !== null) {
            this.controller.unbind();
        }

        bindings = this.bindings;
        for (i = 0, ii = bindings.length; i < ii; ++i) {
            bindings[i].unbind();
        }

        controllers = this.controllers;
        for (i = 0, ii = controllers.length; i < ii; ++i) {
            controllers[i].unbind();
        }

        children = this.children;
        for (i = 0, ii = children.length; i < ii; ++i) {
            children[i].unbind();
        }

        this.bindingContext = null;
        this.overrideContext = null;
    }
};
```

This fixed this [issue](https://github.com/aurelia/binding/issues/361) for me.  But it surfaced another.

Here is the setup:

If I have a custom element setup like this:

    <div repeat.for="box of boxes">
      <box box.bind="box" view-model.ref="boxViewModels[$index]"></box>
    </div>

I then add entries to the `boxes` array in my view-model:

    this.boxes.push(newBox);

This all works great.  But when I remove a box from the array, things get messy.

Say I add 3 boxes.  My `this.boxViewModels` array will look like this:

<pre>
this.boxViewModels
     |
     +-> boxVm1
     +-> boxVm2
     +-> boxVm3
</pre>

And `this.boxes` will will look like this:

<pre>
this.boxes
     |
     +-> box1
     +-> box2
     +-> box3
</pre>

If I then remove `box2` from the this.boxes array, the updated arrays look like this:

<pre>
this.boxViewModels
     |
     +-> boxVm1
     +-> null
     +-> boxVm3
</pre>

<pre>
this.boxes
     |
     +-> box1
     +-> box3
</pre>
         
The difference here is that `this.boxViewModels` sets the second element to null, where `this.boxes` removes the element.

This gets really bad when I add another entry to `this.boxes`.  It ends up looking like this:

<pre>
this.boxViewModels
     |
     +-> boxVm1
     +-> null
     +-> boxVm4
</pre>

<pre>
this.boxes
     |
     +-> box1
     +-> box3
     +-> box4
</pre>
         
Now `this.boxViewModels` does not even have a reference to boxVm3 anymore.

**Expected:**  
When an item is removed from a `repeat-for` array, the corresponding item in the array that is binding to `view-model.ref` it will be removed, not set to null.

**Observed**  
When an item is removed from a `repeat-for` array, the corresponding item in the array that is binding to `view-model.ref` is set to null.

NOTE: I tried to create a gist run for this.  But I a typescript guy and getting it running in just JavaScript was not working (and taking a very long time).  ([Here](https://gist.run/?id=8c16d89bce2fb81f97ce&sha=592d9948dadf68ba5e1057632178efd49492649b) is what I got working in case you want to look.)

I am willing to try again if this is not easily reproducible for you.

**Comment by EisenbergEffect** [comment]
@jdanyow I'm starting to think we made a huge mistake and should never have enabled binding expressions for the ref attribute. Can you see any easy way to solve this? I had no idea people were going to try to create arrays of refs and synchronize those back to an array on the view model.

Should we just say that syncing of arrays isn't supported and explain to developers that overuse of ref would be considered a bad practice because it creates extensive coupling between the view and view model?

**Comment by Vaccano** [comment]
@EisenbergEffect - I don't want to engage in bad practice.

My scenario is that I have some methods on the view-model children that I want to be a callback to the view-model parent. (Things like focusNextChild, isDuplicateChild etc.)   This feature was an easy way to do that.

If there is another way to do that I am fine not using this feature.

**Comment by EisenbergEffect** [comment]
You could just use DOM events. Also, perhaps the `@children('selector')` decorator would work?

**Comment by stoffeastrom** [comment]
Or use a `@bindable`? If you have many you could use use a `@bindable api` where `api` is an object from the parent with needed functions?

**Comment by Vaccano** [comment]
@EisenbergEffect - `@children` looks promising.  But the docs are thin on that one.  Is there any example code out there for that?  (I tried it and got an empty array)

@stoffeastrom - That could work. I will give it a try.

**Comment by Vaccano** [comment]
@EisenbergEffect I was able to work around needing this feature using the `@bindable` method (thanks @stoffeastrom !)

Still does not feel any different to me from a best practices point of view.  I am still keeping a list of the child view-models (so I can know which one is next and previous (for focusing)).  The only difference is that I am maintaining the list now (rather than Aurelia).

However, best practice aside, if getting those extra nulls out of the bound array is not doable, then I, for one, can live with the alternatives.

**Comment by EisenbergEffect** [comment]
It is probably "do-able" however it's highly non-trivial and I'm thinking we might end up with a strange amount of complexity introduced to do it. I'm not sure if it's worth it. @jdanyow may have a better fell for things though. He's our binding guru :)

**Comment by EisenbergEffect** [comment]
@Vaccano If your refs were to view models, that's not such a big issue. I do get concerned when building VMs that have lots of coupling to HTML elements themselves.

**Comment by jdanyow** [comment]
I agree- would be messy/complex to add further support for this scenario in binding.

@Vaccano if the child element is tightly coupled to the parent element, you can express this as a dependency in your Box class.  Let's assume the parent element that uses the boxes is named `BoxParent`, and box-parent.html looks like this:

```html
<template>
  <div repeat.for="box of boxes">
    <box box.bind="box" view-model.ref="boxViewModels[$index]"></box>
   </div>
</template>
```

If you make your Box class look like this:

```js
@inject(BoxParent)
export class Box {
  constructor(parent) {
    this.parent = parent;
  }

  bind() {
    this.parent.addBox(this);
  }

  unbind() {
    this.parent.removeBox(this);
  }
}
```

You can access the parent view model using dependency injection and give it the box instances, etc.  Maybe this will be a more natural fit for you.

**Comment by Vaccano** [comment]
@jdanyow That ended up working very well for me.  Thank you!  

However when I added `bind` to my box view model, my `@bindable` `changeHandler` callbacks were no longer called by Aurelia.  

Not sure if this is intended, but I wrote up an issue and repro for it here: https://github.com/aurelia/binding/issues/363

Either way, I no longer need this issue fixed.  Not sure if you want to close it or leave it be...
